segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,366 @@
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(ex instanceof NoDataAvailableException)) {
+                            log.warn("Processing iteration failed for EventProcessor {}, retrying.", getName(), ex);
+                            lastIterationLatency.set(-1); // Do not report latency, as there has been a failure.
+                        }
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> readEvents() {
+            // Need to get the start and end offsets to properly create async read request.
+            long startOffset = segment.getInfo().getStartOffset();
+            int readLength = (int) Math.min(segment.getInfo().getLength() - startOffset, ProcessorEventSerializer.MAX_TOTAL_EVENT_SIZE);
+            return CompletableFuture.supplyAsync(() -> segment.read(startOffset, readLength, containerOperationTimeout), executor)
+                                    .thenComposeAsync(rr -> AsyncReadResultProcessor.processAll(rr, executor, containerOperationTimeout))
+                                    .thenApply(inputData -> deserializeEvents(inputData, startOffset, readLength));
+        }
+
+        private EventsReadAndTruncationPoints deserializeEvents(BufferView inputData, long startOffset, int maxLength) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long nextOffset = startOffset;
+            final long maxOffset = startOffset + maxLength;
+            BufferView.Reader input = inputData.getBufferViewReader();
+            try {
+                while (nextOffset < maxOffset && events.size() < getConfig().getMaxItemsAtOnce()) {
+                    ProcessorEventData event = ProcessorEventSerializer.deserializeEvent(input);
+                    events.add(event);
+                    // Update the offset to the beginning of the next event.
+                    nextOffset += ProcessorEventSerializer.HEADER_LENGTH + event.getLength();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+                if (events.isEmpty()) {
+                    throw new NoDataAvailableException();
+                }
+            }
+
+            return new EventsReadAndTruncationPoints(events);
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> applyProcessorHandler(EventsReadAndTruncationPoints readResult) {
+            return CompletableFuture.supplyAsync(() -> getHandler().apply(readResult.getProcessorEventsData()), executor)
```

Comment:
Why do you wrap this in a Future? There is overhead in doing so, so if we don't have to, then I'd prefer we didn't.

Reply:
Removed the creation of that future.